ambry-account/src/main/java/com/github/ambry/account/mysql/DatasetDao.java

```diff
@@ -961,36 +994,88 @@ private long generateAutoIncrementedVersionBasedOnSchema(long versionValue, Data
+          throw new IllegalArgumentException(
+              "The current version string " + version + " is not supported for Semantic Long version schema");
+        }
+        long majorLongVersion = versionValue / 1000000000L;
+        long minorLongVersion = (versionValue % 1000000000L) / 1000000L;
+        long patchLongVersion = (versionValue % 1000000L) / 1000L;
+        long revisionLongVersion = versionValue % 1000L;
+        if (MAJOR.equals(version)) {
+          majorLongVersion += 1;
+          minorLongVersion = 0;
+          patchLongVersion = 0;
+          revisionLongVersion = 0;
+          if (majorLongVersion > MAX_SEMANTIC_VERSION_PART_VALUE) {
+            throw new IllegalArgumentException("The largest version for Semantic Major version should be less than "
+                + MAX_SEMANTIC_VERSION_PART_VALUE + ", currentMajorVersion: " + majorLongVersion);
+          }
+        } else if (MINOR.equals(version)) {
+          minorLongVersion += 1;
+          patchLongVersion = 0;
+          revisionLongVersion = 0;
+          if (minorLongVersion > MAX_SEMANTIC_VERSION_PART_VALUE) {
+            throw new IllegalArgumentException("The largest version for Semantic Minor version should be less than "
+                + MAX_SEMANTIC_VERSION_PART_VALUE + ", currentMinorVersion: " + minorLongVersion);
+          }
+        } else if (PATCH.equals(version)) {
+          patchLongVersion += 1;
+          revisionLongVersion = 0;
+          if (patchLongVersion > MAX_SEMANTIC_VERSION_PART_VALUE) {
+            throw new IllegalArgumentException("The largest version for Semantic Patch version should be less than "
+                + MAX_SEMANTIC_VERSION_PART_VALUE + ", currentPatchVersion: " + patchLongVersion);
+          }
+        } else {
+          revisionLongVersion += 1;
+          if (revisionLongVersion > MAX_SEMANTIC_VERSION_PART_VALUE) {
+            throw new IllegalArgumentException("The largest version for Semantic Revision version should be less than "
+                + MAX_SEMANTIC_VERSION_PART_VALUE + ", currentRevisionVersion: " + revisionLongVersion);
+          }
+        }
+        return majorLongVersion * 1000000000 + minorLongVersion * 1000000 + patchLongVersion * 1000 + revisionLongVersion;
       default:
         throw new IllegalArgumentException("Unsupported version schema: " + versionSchema);
     }
   }
 
   /**
    * @param version the version string.
+   * @param versionSchema the {@link com.github.ambry.account.Dataset.VersionSchema} from dataset level.
    * @return {@code} true if the version is auto created version for upload.
    */
-  private boolean isAutoCreatedVersionForUpload(String version) {
-    return LATEST.equals(version) || MAJOR.equals(version) || MINOR.equals(version) || PATCH.equals(version);
+  private boolean isAutoCreatedVersionForUpload(String version, Dataset.VersionSchema versionSchema) {
```

Comment:
I think we can simply return
LATEST.equals(version) || MAJOR.equals(version) || MINOR.equals(version) || PATCH.equals(version) || REVISION.equals(version).
Because if user provide MAJOR, it means auto incr the major version. I think similar thing should be applied for REVISION?